File: codes/golang/303.区域和检索-数组不可变.go


type NumArray struct {
	preSumMap map[int]int
}

func Constructor(nums []int) NumArray {
	preSumMap := make(map[int]int)
	preSum := 0
	for i := 0; i < len(nums); i++ {
		preSum += nums[i]
		preSumMap[i] = preSum
	}
	return NumArray{preSumMap: preSumMap}
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.preSumMap[right] - this.preSumMap[left-1]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(left,right);
 */

File: codes/python/303.区域和检索-数组不可变.py


class NumArray:
	def __init__(self, nums: List[int]):
		self.preSumMap = {}
		preSum = 0
		for i in range(len(nums)):
			preSum += nums[i]
			self.preSumMap[i] = preSum


	def sumRange(self, left: int, right: int) -> int:
		return self.preSumMap.get(right, 0) - self.preSumMap.get(left - 1, 0)



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)